### DISCUSSION TITLE
How to boost perfomance on M-Macs

### DISCUSSION THREAD

**Discussion by ilgrandeanonimo** [lllyasviel/Fooocus#3447] [Show and tell]
These days I was playing around with Fooocus, I've tested some of the commands from https://github.com/lllyasviel/Fooocus/pull/129 to find which one will generate stuff faster.

Here the results of my tests:
🐢 Slowest
`python entry_with_update.py --attention-pytorch` 
this is the worst, 120s/it, 60s to load model

🐱Nothing special
`python entry_with_update.py --always-cpu --disable-offload-from-vram --unet-in-fp8-e5m2` 
25-20s/it, 45s to load model

🐆 Fast
`python entry_with_update.py --unet-in-fp16 --attention-split` 
initially 20 then 14-15s/it, 40s to load model

🌟Very Fast
`python entry_with_update.py --all-in-fp16 --attention-pytorch --disable-offload-from-vram --always-high-vram --gpu-device-id 0 --unet-in-fp16 --vae-in-fp16 --clip-in-fp16` 
This is the best, 10s/it and only 24s to load the model

**My System:**
MacBook Air 13", M1 2020, 16GB, 8-Core CPU & 8 Core GPU with macOS Sonoma 14.6

**Other informations**
Ambient Temperature: 35°C
To prevent my Mac from melting I've put it on some "ice bricks". I would recommend finding a better solution to cool a computer if you need to.

**ATTENTION**: Using it on a Mac with not enough RAM can cause massive SWAP usage and drain the SSD's lifespan.

**Comment by mashb1t** [comment]
It's a Macbook Air and you're already getting good results for what it is using the MPS backend. Used libraries have already been optimized in 2.5.0, further optimization requires https://github.com/lllyasviel/Fooocus/pull/3084, but i haven't yet received any support on this.

**Reply by ilgrandeanonimo** [reply]
> It's a Macbook Air and you're already getting good results for what it is using the MPS backend. Used libraries have already been optimized in 2.5.0, further optimization requires #3084, but i haven't yet received any support on this.

Sorry, maybe I posted in the wrong place. It was more of a post where I wanted to inform other users how to boost performance on mac.

**Reply by mashb1t** [reply]
@ilgrandeanonimo got it, i've adjusted the category to "show & tell" and changed the title for increased clarity.

**Comment by shotexa** [comment]
`--disable-offload-from-vram` actually makes my m1 mac slower, I'm on 16 inch m1 max 2021

**Reply by ilgrandeanonimo** [reply]
> --disable-offload-from-vram actually makes my m1 mac slower, I'm on 16 inch m1 max 2021

I tried to remove `--disable-offload-from-vraw` and in my case performance decrased:
model loading from ~24s to ~42s and the image generation from 10it/s to 15it/s. 

It probably depends on the mac model or some different setting

**Comment by lgwjames** [comment]
on my intel Mac I was getting 14it 900 to 1400 seconds a image. at that was the best I could get. I have just done a egpu set up at a cost of £350 and now under 1min a image

**Reply by zomboy007** [reply]
I got an Intel Mac too and have been trying to figure out how to speed up the generations. 

This is the first time I have heard of an EGPU. 

Which ones would you recommend?

**Reply by lgwjames** [reply]
well firstly 2 things, some wont work with imac, the gtx3060 12gb i have cant be seen. looks like you need amd ones as per guide. 
The one i have can be seen by bootcamp on win11. 1  problem, i have no sound from mac speakers. tried everything, i get sound and then reboot no sound. but here its right now times.

Image generated with private log at: C:\Users\jessi\Desktop\Fooocus_win64_2-5-0\Fooocus\outputs\2024-08-09\log.html
Generating and saving time: 53.60 seconds
[Fooocus] Preparing task 3/5 ...
[Sampler] refiner_swap_method = joint
[Sampler] sigma_min = 0.0291671771556139, sigma_max = 14.614643096923828
Requested to load SDXL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 4.09 seconds
100%|██████████████████████████████████████████████████████████████████████████████| 60/60 [00:48<00:00,  1.25it/s]
Requested to load AutoencoderKL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 0.50 seconds
[Fooocus] Saving image 3/5 to system ...
Image generated with private log at: C:\Users\jessi\Desktop\Fooocus_win64_2-5-0\Fooocus\outputs\2024-08-09\log.html
Generating and saving time: 53.76 seconds
[Fooocus] Preparing task 4/5 ...
[Sampler] refiner_swap_method = joint
[Sampler] sigma_min = 0.0291671771556139, sigma_max = 14.614643096923828
Requested to load SDXL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 4.04 seconds
100%|██████████████████████████████████████████████████████████████████████████████| 60/60 [00:47<00:00,  1.25it/s]
Requested to load AutoencoderKL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 0.49 seconds
[Fooocus] Saving image 4/5 to system ...
Image generated with private log at: C:\Users\jessi\Desktop\Fooocus_win64_2-5-0\Fooocus\outputs\2024-08-09\log.html
Generating and saving time: 53.60 seconds
[Fooocus] Preparing task 5/5 ...
[Sampler] refiner_swap_method = joint
[Sampler] sigma_min = 0.0291671771556139, sigma_max = 14.614643096923828
Requested to load SDXL
Loading 1 new model
[Fooocus Model Management] Moving model(s) has taken 4.07 seconds
100%|██████████████████████████████████████████████████████████████████████████████| 60/60 [00:47<00:00,  1.25it/s]
